src/ObjectUtils/CoffFile.cpp

```diff
@@ -38,82 +39,58 @@ class CoffFileImpl : public CoffFile {
-  has_symbol_table_ = (object_file_->getSymbolTable() != 0);
-}
-
-ErrorMessageOr<uint64_t> CoffFileImpl::GetSectionOffsetForSymbol(
-    const llvm::object::SymbolRef& symbol_ref) {
-  // Symbol addresses are relative virtual addresses (RVAs) that are relative to the
-  // start of the section. To get the address relative to the start of the object
-  // file, we need to find the section and add the base address of the section.
-  llvm::object::COFFSymbolRef coff_symbol_ref = object_file_->getCOFFSymbol(symbol_ref);
-  int32_t section_id = coff_symbol_ref.getSectionNumber();
-  const llvm::object::coff_section* section = nullptr;
-  std::error_code err = object_file_->getSection(section_id, section);
-  if (err) {
-    return ErrorMessage(absl::StrFormat("Section of symbol not found: %s", err.message()));
+  const auto dwarf_context = llvm::DWARFContext::create(*owning_binary_.getBinary());
+  if (dwarf_context != nullptr) {
+    has_dwarf_debug_info_ = true;
   }
-  return section->VirtualAddress;
 }
 
-ErrorMessageOr<SymbolInfo> CoffFileImpl::CreateSymbolInfo(
-    const llvm::object::SymbolRef& symbol_ref) {
-  if ((symbol_ref.getFlags() & llvm::object::BasicSymbolRef::SF_Undefined) != 0) {
-    return ErrorMessage("Symbol is defined in another object file (SF_Undefined flag is set).");
-  }
-  const std::string name = symbol_ref.getName() ? symbol_ref.getName().get() : "";
-  // Unknown type - skip and generate a warning.
-  if (!symbol_ref.getType()) {
-    LOG("WARNING: Type is not set for symbol \"%s\" in \"%s\", skipping.", name,
-        file_path_.string());
-    return ErrorMessage(absl::StrFormat(R"(Type is not set for symbol "%s" in "%s", skipping.)",
-                                        name, file_path_.string()));
-  }
-  // Limit list of symbols to functions. Ignore sections and variables.
-  if (symbol_ref.getType().get() != llvm::object::SymbolRef::ST_Function) {
-    return ErrorMessage("Symbol is not a function.");
-  }
-
-  auto section_offset = GetSectionOffsetForSymbol(symbol_ref);
-  if (section_offset.has_error()) {
-    return ErrorMessage(absl::StrFormat("Error retrieving section offset for symbol \"%s\": %s",
-                                        name, section_offset.error().message()));
+static void FillDebugSymbolsFromDWARF(uint64_t image_base, llvm::DWARFContext* dwarf_context,
+                                      ModuleSymbols* module_symbols) {
+  for (const auto& info_section : dwarf_context->compile_units()) {
+    for (uint32_t index = 0; index < info_section->getNumDIEs(); ++index) {
+      llvm::DWARFDie full_die = info_section->getDIEAtIndex(index);
+      // We only want symbols of functions, which are DIEs with isSubprogramDIR()
+      // and not of inlined functions, which are DIEs with isSubroutineDIE().
+      if (full_die.isSubprogramDIE()) {
+        SymbolInfo symbol_info;
+        std::string name(full_die.getName(llvm::DINameKind::LinkageName));
```

Comment:
This can present a problem actually. The linkage name exists only if it is different from DW_AT_NAME. So this could be empty in which case we should probably fall back to DW_AT_NAME.

And of course there is this, to make things more complicated: https://bugs.llvm.org/show_bug.cgi?id=37470

Reply:
Ok, added fallback to DW_AT_NAME. Maybe it's also okay for us to just always use the short name?

Reply:
> Maybe it's also okay for us to just always use the short name?
Then we can run into the bug with removed underscore. It linkage_name is more reliable if present.

Reply:
As discussed, the llvm getName method actually falls back to ShortName if LinkageName is found. I added a CHECK against an empty string here, which should never happen.